### PR TITLE
bug #4402 - showing failed transactions in transaction history

### LINKED ISSUE #4402: Show failed transactions in transaction history as failed
### User Story

As a user I want clear information about transactions I made

### Description

*Type*: Bug

*Summary*: Failed transactions should be shown in transaction history as Failed

#### Expected behavior
https://ropsten.etherscan.io/tx/0x716eefb91b2d33f63687222f2db2125813098f61fb744d9ea0878a3c45e239ed should be shown as Failed so i am clear enough what has happened with my funds

#### Actual behavior
Failed transactions are not marked as such so i am expecting them to be processed normally when it is basically not

<blockquote><div><strong><a href="https://ropsten.etherscan.io/tx/0x716eefb91b2d33f63687222f2db2125813098f61fb744d9ea0878a3c45e239ed">Ropsten Transaction 0x716eefb91b2d33f63687222f2db2125813098f61fb744d9ea0878a3c45e239ed</a></strong></div><div>Ropsten (ETH) detailed transaction info for 0x716eefb91b2d33f63687222f2db2125813098f61fb744d9ea0878a3c45e239ed</div></blockquote>

### LINKED ISSUE #4639: "BigNumber Error: div() not a number: null" in transaction history if user has failed transactions
### Description
*Type*: Bug

**Summary:** 

**If user has failed transactions in history - after recovering account he cannot use transaction history (error will appear ~every minute)**

*Scenario 1:*
Can be reproducible for failed transaction, i.e. for sending to contract address that cannot be executed (contract example - `0xb3fe697300ec8d3777f2d9a8b140f6f9f5df665b`)
TX hash: https://etherscan.io/tx/0x4e4855207cc1a1a418bb640dc2ec3e83c8815d3a869698d685f1b68a656f4e10.

*Scenario 2:*
when you are trying to send small ETH amount (in my case 0,0000001 ETH) with minimal non-divisible gas Price (0,000000001 GWei) I don't see any errors, but eventually transaction will be rejected by network and won't appear. After a while instead of confirmations  "BigNumber Error: div() not a number: null" appears in Transaction History and transaction itself disappears out of "Transaction History" screen
tr_h

*Scenario 3:*
restore account of user who has failed transactions in history (like this https://etherscan.io/address/0xcf2272205cc0cf96cfbb9dd740bd681d1e86901e)

#### Expected behavior
*Scenario 1:*
transaction is failed 
*Scenario 2:*
Transaction is in pending state; i.e. in MetaMask with the same data transaction is Pending and you can resend it with higher Gas Price
pend_2
*Scenario 3:*
no error

#### Actual behavior
"BigNumber Error: div() not a number: null" error

### Reproduction
*Network:* Mainnet
*Scenario 1:*
- Open Status
- Open Wallet
- Enter unexecutable contract address (i.e. `0xb3fe697300ec8d3777f2d9a8b140f6f9f5df665b`) to emulate failed transactions
- Enter 0,00001 ETH in amount
- Tap on "Send transaction"
- Go to Transaction history and wait ~1 min

*Scenario 2:*
- Open Status
- Open Wallet
- Enter valid address
- Enter 0,0000001 ETH in amount
- Tap on Advanced
- Specify 0.000000001 GWei as Gas Price
- Tap on "Done"
- Tap on "Send transaction"
- Go to Transaction history and wait ~1 min

*Scenario 3:*
- Open Status
- Restore account with Failed transactions (i.e. https://etherscan.io/address/0xcf2272205cc0cf96cfbb9dd740bd681d1e86901e)
- Go to Transaction history and wait ~1 min

### Additional Information
* Status version: nightly 06/06/2018
* Operating System: Android, iOS

#### Logs
* TF sesson (scenario 3): https://app.testfairy.com/projects/4803590-status/builds/8451016/sessions/14/?accessToken=KWNCc1z90oXlwBWLrRoEl0s-UGc

### PR DESCRIPTION
**Pull request by goranjovic**
fixes #4402
fixes #4639

### Summary: 

Clearly marks failed transactions as such. Does so by parsing a "failed" flag provided by etherscan.

EDIT:

Also fixes #4639 - the problem here was that exchanges sometimes emit transfer events for internal contracts they use alongside the transfer events for tokens that we want to show. 

### Steps to test:

see #4639

status: ready

### CODE REVIEW

**Comment by goranjovic** [comment]
screenshot-2018-06-09_13 31 18 908
----
screenshot-2018-06-09_12 42 38 582

**Comment by goranjovic** [comment]
Could not reproduce `BigNumber error div()` scenario. For other failing transactions, showing them as failed as on screenshots.

**Review by jeluard — APPROVED** [review]

**Comment by denis-sharypin** [comment]
@goranjovic  hiii, the color of the values should be black
screenshot 2018-06-12 11 04 44

**Comment by denis-sharypin** [comment]
Please use color of `failed` for exclamation icon 

screenshot 2018-06-12 11 06 25

**Comment by goranjovic** [comment]
screenshot-2018-06-12_12 57 35 666

Updated value amount a black color. 
@denis-sharypin The exclamation mark is already using the same color as the "Failed" text, not sure why it looks different

File: src/status_im/utils/transactions.cljs [hunk 1]
```diff
@@ -25,23 +25,24 @@
...
-     :hash  hash
-     :from from
-     :to to
-     :gas-limit gas
-     :gas-price gasPrice
-     :gas-used gasUsed
-     :nonce nonce
+     :timestamp     (str timeStamp "000")
+     :symbol        :ETH
+     :type          (if error? :failed (if inbound? :inbound :outbound))
```
Comment (rasom) [inline src/status_im/utils/transactions.cljs]: `(cond error? :failed inbound? :inbound :else :outbound)` `cond` would work here, though the only advantage - less parentheses (if that's advantage for lisp)

Comment (goranjovic) [reply src/status_im/utils/transactions.cljs]: It's certainly cleaner, for sure. 

https://imgs.xkcd.com/comics/lisp_cycles.png

Comment (goranjovic) [reply src/status_im/utils/transactions.cljs]: @rasom updated

**Review by mandrigin — APPROVED** [review]
looks good!

**Comment by asemiankevich** [comment]
looks good to me, no errors . lets merge

**Pull request merged**